File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/examples/BatchExportExample.java
package com.pb40.bimportal.examples;

import com.bimportal.client.model.SimpleAiaProjectPublicDto;
import com.bimportal.client.model.SimpleLoinPublicDto;
import com.pb40.bimportal.client.BimPortalClientBuilder;
import com.pb40.bimportal.client.EnhancedBimPortalClient;
import com.pb40.bimportal.config.BimPortalConfig;
import com.pb40.bimportal.util.ExportUtils;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Batch export examples for BIM Portal Java client.
 *
 * <p>Demonstrates batch export workflows for multiple projects and LOINs in PDF format for
 * efficient bulk processing.
 */
public class BatchExportExample {

  private static final Logger logger = LoggerFactory.getLogger(BatchExportExample.class);

  /**
   * Check if credentials are available for authentication.
   *
   * @return True if credentials are configured
   */
  private static boolean checkCredentials() {
    if (!BimPortalConfig.hasCredentials()) {
      System.out.println("============================================================");
      System.out.println("WARNING: Credentials not found in environment variables.");
      System.out.println(
          "Please set "
              + BimPortalConfig.USERNAME_ENV_VAR
              + " and "
              + BimPortalConfig.PASSWORD_ENV_VAR
              + " in .env file.");
      System.out.println("============================================================");
      return false;
    }
    return true;
  }

  /**
   * Run batch export examples for projects and LOINs.
   *
   * @param client Enhanced BIM Portal client
   */
  public static void runBatchExportExamples(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("📦 BATCH EXPORT EXAMPLES");
    System.out.println("============================================================");

    int totalExported = 0;

    System.out.println("\n1️⃣ Batch exporting projects as PDF...");
    try {
      List<SimpleAiaProjectPublicDto> projects = client.searchProjects();
      if (!projects.isEmpty()) {
        int projectCount = Math.min(3, projects.size());
        for (int i = 0; i < projectCount; i++) {
          SimpleAiaProjectPublicDto project = projects.get(i);
          System.out.println(
              "   🚧 Exporting project " + (i + 1) + "/" + projectCount + ": " + project.getName());

          Optional<byte[]> pdfContent = client.exportProjectPdf(project.getGuid());
          if (pdfContent.isPresent()) {
            String filename = "batch_project_" + (i + 1) + "_" + project.getGuid() + ".pdf";
            Optional<Path> savedPath = ExportUtils.saveExportFile(pdfContent.get(), filename);
            if (savedPath.isPresent()) {
              totalExported++;
              System.out.println("   ✅ Success: " + savedPath.get());
            } else {
              System.out.println("   ❌ Failed: File save error");
            }
          } else {
            System.out.println("   ❌ Failed: " + project.getName());
          }
        }
      } else {
        System.out.println("🔭 No projects available");
      }
    } catch (Exception e) {
      logger.error("❌ Error in project batch export", e);
      System.err.println("❌ Error in project batch export: " + e.getMessage());
    }

    System.out.println("\n2️⃣ Batch exporting LOINs as PDF...");
    try {
      List<SimpleLoinPublicDto> loins = client.searchLoins();
      if (!loins.isEmpty()) {
        int loinCount = Math.min(2, loins.size());
        for (int i = 0; i < loinCount; i++) {
          SimpleLoinPublicDto loin = loins.get(i);
          System.out.println(
              "   📊 Exporting LOIN " + (i + 1) + "/" + loinCount + ": " + loin.getName());

          Optional<byte[]> pdfContent = client.exportLoinPdf(loin.getGuid());
          if (pdfContent.isPresent()) {
            String filename = "batch_loin_" + (i + 1) + "_" + loin.getGuid() + ".pdf";
            Optional<Path> savedPath = ExportUtils.saveExportFile(pdfContent.get(), filename);
            if (savedPath.isPresent()) {
              totalExported++;
              System.out.println("   ✅ Success: " + savedPath.get());
            } else {
              System.out.println("   ❌ Failed: File save error");
            }
          } else {
            System.out.println("   ❌ Failed: " + loin.getName());
          }
        }
      } else {
        System.out.println("🔭 No LOINs available");
      }
    } catch (Exception e) {
      logger.error("❌ Error in LOIN batch export", e);
      System.err.println("❌ Error in LOIN batch export: " + e.getMessage());
    }

    System.out.println(
        "\n📊 BATCH EXPORT COMPLETE: " + totalExported + " files exported successfully");
  }

  /** Main method to run batch export examples. */
  public static void main(String[] args) {
    System.out.println("======================================================================");
    System.out.println("🚀 BIM PORTAL BATCH EXPORT EXAMPLES");
    System.out.println("======================================================================");

    if (!checkCredentials()) {
      System.err.println("❌ Cannot run export examples without credentials");
      return;
    }

    if (!ExportUtils.isExportDirectoryWritable()) {
      System.err.println("❌ Export directory is not writable!");
      return;
    }

    System.out.println("🔧 Setting up authenticated client...");

    try {
      EnhancedBimPortalClient client = BimPortalClientBuilder.buildDefault();

      EnhancedBimPortalClient.HealthCheckResult healthCheck = client.performHealthCheck();
      System.out.println("🩺 Health check: " + healthCheck);

      if (!healthCheck.isApiAccessible()) {
        System.err.println("❌ API is not accessible. Please check configuration.");
        return;
      }

      runBatchExportExamples(client);

      System.out.println(
          "\n======================================================================");
      System.out.println("✅ BATCH EXPORT EXAMPLES COMPLETE!");
      System.out.println("======================================================================");
      System.out.println(
          "📂 Check the '" + BimPortalConfig.getExportPath() + "' directory for exported files");

      client.logout();
      System.out.println("👋 Logged out and session closed.");

    } catch (Exception e) {
      logger.error("❌ Error running batch export examples", e);
      System.err.println("❌ Error running examples: " + e.getMessage());
    }
  }
}

File: Python_client/examples/export_examples/batch_export_example.py
"""
Batch export examples for BIM Portal Python client (Refactored).

This module demonstrates batch export workflows for multiple projects and LOINs
in PDF format for efficient bulk processing.
Refactored to use simple common utility functions.
"""
import os
import sys
from pathlib import Path

# Ensure we can import from project root
project_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(project_root))

import logging
from examples.export_examples.utils.common_utils import (
    handle_main_example_setup,
    print_example_footer,
    print_section_header,
    export_single_item_batch,
    run_cleanup_operations
)

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def run_batch_export_examples(client):
    """
    Run batch export examples for projects and LOINs.

    Args:
        client: Enhanced BIM Portal client
    """
    print_section_header("BATCH EXPORT EXAMPLES")

    total_exported = 0

    print("\n1️⃣ Batch exporting projects as PDF...")
    try:
        projects = client.search_projects()
        if projects:
            project_count = min(3, len(projects))
            successful_projects = 0

            for i, project in enumerate(projects[:project_count]):
                success = export_single_item_batch(
                    client, project, "project",
                    client.export_project_pdf, i+1, project_count
                )
                if success:
                    successful_projects += 1

            total_exported += successful_projects
            print(f"   📊 Projects: {successful_projects}/{project_count} exported successfully")
        else:
            print("🔭 No projects available")

    except Exception as e:
        logger.error("Error in project batch export", exc_info=True)
        print(f"❌ Error in project batch export: {e}")

    print("\n2️⃣ Batch exporting LOINs as PDF...")
    try:
        loins = client.search_loins()
        if loins:
            loin_count = min(2, len(loins))
            successful_loins = 0

            for i, loin in enumerate(loins[:loin_count]):
                success = export_single_item_batch(
                    client, loin, "loin",
                    client.export_loin_pdf, i+1, loin_count
                )
                if success:
                    successful_loins += 1

            total_exported += successful_loins
            print(f"   📊 LOINs: {successful_loins}/{loin_count} exported successfully")
        else:
            print("🔭 No LOINs available")

    except Exception as e:
        logger.error("Error in LOIN batch export", exc_info=True)
        print(f"❌ Error in LOIN batch export: {e}")

    print(f"\n📊 BATCH EXPORT COMPLETE: {total_exported} files exported successfully")

    # Additional batch operations
    if total_exported > 0:
        print("\n3️⃣ Additional batch operations...")
        run_cleanup_operations()


def run_advanced_batch_operations(client):
    """
    Demonstrate advanced batch operations with error handling and progress tracking.

    Args:
        client: Enhanced BIM Portal client
    """
    print("\n4️⃣ Advanced batch operations with error handling...")

    successful_exports = 0
    failed_exports = 0

    try:
        # Get resources for batch processing
        projects = client.search_projects()
        loins = client.search_loins()

        # Process up to 2 projects
        project_count = min(2, len(projects))
        print(f"   📄 Processing {project_count} projects...")

        for i, project in enumerate(projects[:project_count]):
            success = export_single_item_batch(
                client, project, "project",
                client.export_project_pdf, i+1, project_count
            )
            if success:
                successful_exports += 1
            else:
                failed_exports += 1

        # Process up to 2 LOINs
        loin_count = min(2, len(loins))
        print(f"   📋 Processing {loin_count} LOINs...")

        for i, loin in enumerate(loins[:loin_count]):
            success = export_single_item_batch(
                client, loin, "loin",
                client.export_loin_pdf, i+1, loin_count
            )
            if success:
                successful_exports += 1
            else:
                failed_exports += 1

        # Print summary
        total_processed = successful_exports + failed_exports
        print(f"\n   📊 Advanced Batch Summary:")
        print(f"      ✅ Successful: {successful_exports}")
        print(f"      ❌ Failed: {failed_exports}")
        print(f"      📈 Total processed: {total_processed}")

    except Exception as e:
        logger.error("Error in advanced batch operations", exc_info=True)
        print(f"   ❌ Error in advanced batch operations: {e}")


def main():
    """Main method to run batch export examples."""
    client = handle_main_example_setup("BATCH")
    if not client:
        return

    try:
        # Run basic batch export examples
        run_batch_export_examples(client)

        # Run advanced batch operations
        run_advanced_batch_operations(client)

        print_example_footer("BATCH")

    except Exception as e:
        logger.error("Error running batch export examples", exc_info=True)
        print(f"❌ Error running examples: {e}")


if __name__ == "__main__":
    main()